rfcs/plugin-api/experiment-2.ts

```diff
@@ -0,0 +1,45 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable @typescript-eslint/no-empty-function */
+import { Config, EnvironmentFunctions } from "@inlang/core/src/config"
+
+export const myPlugin = createPlugin<{ pathPattern: string }>(({ pluginConfig, env }) => {
+	return {
+		id: "samuelstroschein.plugin-json",
+		defineConfig: (config) => {
+			if (pluginConfig.pathPattern === undefined) {
+				throw new Error("pathPattern is required")
+			}
+			config.readResources = readResources({ pluginConfig, env })
+			config.languages = getLanguages({ pluginConfig, env })
```

Comment:
We will probably regret mutating objects in the future. I think `defineConfig` should return a new copy of the config. See `experiment-3.ts`

Reply:
Returning a new config could definitely make sense!! What speaks against a new copy of the config and for mutating?

Reply:
**Reasons against mutable config**

- requires additional logic/takes longer to implement because a merging logic needs to be implemented. @ivanhofer spreading will break because it's shallow.
- vite has a mutable and immutable config definition because mutable is required when merging logic fails.
- (performance hit. maybe important in the browser because) 

**But, we can have a simple trick to path the way for immutable config definition in the future**

We can enforce a return value. By forcing a return value, we can implement immutability at a later point by passing a copied `config` object into `defineConfig`. 

```diff
type Plugin = {
	id: string
-	defineConfig(config: Partial<Config>): void
+	defineConfig(config: Partial<Config>): Partial<Config>

}
```

```diff
export const myPlugin = createPlugin<{ pathPattern: string }>(({ pluginConfig, env }) => {
	return {
		id: "samuelstroschein.plugin-json",
		defineConfig: (config) => {
			if (pluginConfig.pathPattern === undefined) {
				throw new Error("pathPattern is required")
			}
			config.readResources = readResources({ pluginConfig, env })
			config.languages = getLanguages({ pluginConfig, env })
+			return config
		},
	}
})
```

**How Vite defines the Plugin API**

Vite uses two separate callbacks which we have in one. 

1. `config` to define the config. 
2. `resolvedConfig` to "validate" the resulting config of all plugins. 

I applied Vite's design to our API:

```ts
type Plugin = {
	id: string
	config: () => Partial<Config>
	resolvedConfig: (config: Config): void
}
```


```ts
export const myPlugin =
  createPlugin <
  { pathPattern: string } >
  (({ pluginConfig, env }) => {
    return {
      id: "samuelstroschein.plugin-json",
      config: () => {
        if (pluginConfig.pathPattern === undefined) {
          throw new Error("pathPattern is required");
        }
        return {
          readResources: readResources({ pluginConfig, env }),
          languages: getLanguages({ pluginConfig, env }),
        };
      },
      resolvedConfig: (config) => {
        if (config.ideExtension === undefined) {
          throw Error(
            "config.ideExtension must be defined for this plugin to work."
          );
        }
      },
    };
  });
```

I think vite's design is better for the following reasons: 

- The ordering of plugins matters less, leading to increased DX. The configs of all plugins will be merged together. Thus, if a plugin requires `readResources` to be defined, the plugin that defines `readResources` doesn't have to be defined before the consuming plugin. 
- Developers have to write less code. By us merging the configs together, developers don't have to write code that, for example, checks if the lint array is `undefined` before adding a lint rule. 
- We can have error handling in the merge logic e.g. `"readResources has been defined by a previous plugin"` in the future.

Going forward, I would choose the Vite plugin API with `config` that is immutable and `resolvedConfig` as an option for plugins to validate configs.

Reply:
Agree. Merging the config internally makes sense! I didn't want to suggest this because in the past my "bullet-prove" ways of doing things were rejected in favor of simpler solutions ^^